### PR TITLE
Longhaul - Updated how we collect system metrics

### PR DESCRIPTION
**Pull request by tmahmood-microsoft**
Using implementation from https://github.com/Azure/azure-iot-sdk-csharp/blob/main/e2e/stress/IoTClientPerf/Reporting/SystemMetrics.cs  as performance counters do not work with Linux and performance counters for .NET CLR were not working as expected.

### CODE REVIEW

**Review by patilsnr — APPROVED** [review]

File: e2e/LongHaul/device/SystemHealthTelemetry.cs [hunk 1]
```diff
@@ -1,53 +1,38 @@
...
+            s_tcpPortFilter = port;
+        }
 
-        [JsonPropertyName("processCpuUsagePercent")]
-        public float ProcessCpuUsagePercent { get; set; } = s_processCpuCounter.NextValue();
-
-        [JsonPropertyName("processWorkingSet")]
-        public float ProcessWorkingSet { get; set; } = s_processWorkingSet.NextValue();
+        private static readonly Process s_currentProcess = Process.GetCurrentProcess();
+        public static long s_tcpPortFilter;
```
Comment (brycewang-microsoft) [inline e2e/LongHaul/device/SystemHealthTelemetry.cs:15]: Variables should go before the ctor.

File: e2e/LongHaul/device/SystemHealthTelemetry.cs [hunk 2]
```diff
@@ -1,53 +1,38 @@
...
+        public double processCpu { get; set; } = s_currentProcess.TotalProcessorTime.TotalMilliseconds / Environment.ProcessorCount;
+        public long totalAssignedMemoryBytes { get; set; } = s_currentProcess.WorkingSet64;
+        public long totalGCBytes { get; set; } = GC.GetTotalMemory(false);
+        public long tcpConnections { get; set; } = updateTCPConnections();
```
Comment (brycewang-microsoft) [inline e2e/LongHaul/device/SystemHealthTelemetry.cs:19]: Properties naming.
```suggestion
        public double ProcessCpu { get; set; } = s_currentProcess.TotalProcessorTime.TotalMilliseconds / Environment.ProcessorCount;
        public long TotalAssignedMemoryBytes { get; set; } = s_currentProcess.WorkingSet64;
        public long TotalGCBytes { get; set; } = GC.GetTotalMemory(false);
        public long TcpConnections { get; set; } = updateTCPConnections();
```

File: e2e/LongHaul/device/SystemHealthTelemetry.cs [hunk 3]
```diff
@@ -1,53 +1,38 @@
...
 
-        [JsonPropertyName("processWorkingSetPrivate")]
-        public float ProcessWorkingSetPrivate { get; set; } = s_processWorkingSetPrivate.NextValue();
 
-        [JsonPropertyName("processPrivateBytes")]
-        public float ProcessPrivateBytes { get; set; } = s_processPrivateBytes.NextValue();
+        private static long updateTCPConnections()
```
Comment (brycewang-microsoft) [inline e2e/LongHaul/device/SystemHealthTelemetry.cs:22]: Method naming.

```suggestion
        private static long UpdateTCPConnections()
```

**Review by brycewang-microsoft — APPROVED** [review]

File: e2e/LongHaul/device/SystemHealthTelemetry.cs [hunk 4]
```diff
@@ -1,53 +1,43 @@
...
+        public long TotalGCBytes { get; set; } = GC.GetTotalMemory(false);
+        public long TcpConnections { get; set; } = UpdateTCPConnections();
+
+        public SystemHealthTelemetry(int port)
+        {
+            TcpPortFilter = port;
+        }
+
+
+        private static long UpdateTCPConnections()
```
Comment (drwill-ms) [inline e2e/LongHaul/device/SystemHealthTelemetry.cs:27]:
```suggestion
        private static long UpdateTcpConnections()
```

File: e2e/LongHaul/device/SystemHealthTelemetry.cs [hunk 5]
```diff
@@ -1,53 +1,43 @@
...
-            true);
-        private static readonly PerformanceCounter s_processBytesInAllHeaps = null;
-        //new PerformanceCounter(
-        //    ".NET CLR Memory",
-        //    "# Bytes in all Heaps",
-        //    _processName,
-        //    true);
-
-        [JsonPropertyName("processCpuUsagePercent")]
-        public float ProcessCpuUsagePercent { get; set; } = s_processCpuCounter.NextValue();
```
Comment (drwill-ms) [inline e2e/LongHaul/device/SystemHealthTelemetry.cs:39]: Why not keep the properties and have them initialize to the appropriate method?

File: e2e/LongHaul/device/SystemHealthTelemetry.cs [hunk 6]
```diff
@@ -1,53 +1,43 @@
...
+using System;
+using System.Diagnostics;
+using System.Net.NetworkInformation;
 
 namespace Microsoft.Azure.Devices.LongHaul.Device
 {
     internal class SystemHealthTelemetry : TelemetryBase
     {
+
+
```
Comment (drwill-ms) [inline e2e/LongHaul/device/SystemHealthTelemetry.cs:13]:
```suggestion
```

File: e2e/LongHaul/device/SystemHealthTelemetry.cs [hunk 7]
```diff
@@ -1,53 +1,43 @@
...
+            long n = 0;
+            foreach (TcpConnectionInformation conn in connections)
+            {
+                if (TcpPortFilter != 0 && conn.RemoteEndPoint.Port != TcpPortFilter)
+                    continue;
+                if (conn.State == TcpState.Established)
+                    n++;
+            }
+            return n;
```
Comment (drwill-ms) [inline e2e/LongHaul/device/SystemHealthTelemetry.cs:40]: Try Linq? Something like this:

```suggestion
            return connections
                .Where(c => TcpPortFilter != 0 && c.RemoteEndPoint.Port != TcpPortFilter)
                . Where(c => c.State == TcpState.Established)
                .Count();
```

File: e2e/LongHaul/service/Program.cs [hunk 8]
```diff
@@ -102,6 +104,11 @@ await Task
             s_aiLoggingProvider.Dispose();
         }
 
+        private static int PortFilter(Parameters parameters)
+        {
+            return parameters.TransportProtocol == IotHubTransportProtocol.WebSocket ? 443 : 5671;
```
Comment (drwill-ms) [inline e2e/LongHaul/service/Program.cs:109]: The problem with this is that even if we choose to use TCP or WebSocket here, AMQP operations on the client will still be on a different port.

File: e2e/LongHaul/device/Program.cs [hunk 9]
```diff
@@ -128,6 +133,18 @@ private static Logger InitializeLogging(Parameters parameters)
             return logger;
         }
 
+        private static int PortFilter(Parameters parameters)
+        {
+            return parameters.TransportProtocol == IotHubClientTransportProtocol.WebSocket
```
Comment (drwill-ms) [inline e2e/LongHaul/device/Program.cs:138]: The problem with this is that even if we choose to use AMQP/MQTT and TCP/WebSocket here, HTTP operations on the client will still be on a different port.

**Review by drwill-ms — CHANGES_REQUESTED** [review]
Comments inline

File: e2e/LongHaul/service/Program.cs (hunk 8 above)
Comment (tmahmood-microsoft) [reply e2e/LongHaul/service/Program.cs:109]: @drwill-ms should I remove this implementation or you think there might be a workaround here?

Comment (drwill-ms) [reply e2e/LongHaul/service/Program.cs:109]: Can you filter in more than 1 port?

Comment (tmahmood-microsoft) [reply e2e/LongHaul/service/Program.cs:109]: sure, we can pass in array of ports as a parameter when filtering.

Comment (drwill-ms) [reply e2e/LongHaul/service/Program.cs:109]: Then for the device client, always include 443 for HTTPS, and for the service client, always include 443 for HTTP and 5671 for AMQP.

File: e2e/LongHaul/device/SystemHealthTelemetry.cs [hunk 10]
```diff
@@ -1,53 +1,72 @@
...
+        [JsonPropertyName("totalAssignedMemoryBytes")]
+        public long TotalAssignedMemoryBytes { get; set; } = s_currentProcess.WorkingSet64;
+
+        [JsonPropertyName("totalGCBytes")]
+        public long TotalGCBytes { get; set; } = GC.GetTotalMemory(false);
+
+        [JsonPropertyName("activeTcpConnections")]
+        public long ActiveTcpConnections { get; set; } = UpdateTcpConnections();
+
+        public SystemHealthTelemetry(int port)
```
Comment (drwill-ms) [inline e2e/LongHaul/device/SystemHealthTelemetry.cs:33]: ctor above properties

File: e2e/LongHaul/device/SystemHealthTelemetry.cs [hunk 11]
```diff
@@ -1,53 +1,72 @@
...
+        {
+            TcpPortFilter.Add(port);
+        }
+
+
```
Comment (drwill-ms) [inline e2e/LongHaul/device/SystemHealthTelemetry.cs:38]:
```suggestion

```

**Review by drwill-ms — APPROVED** [review]

File: e2e/LongHaul/device/IotHub.cs [hunk 12]
```diff
@@ -121,10 +124,10 @@ public async Task SendTelemetryMessagesAsync(CancellationToken ct)
                 }
 
                 // If not connected, skip the work below this round
-                if (!IsConnected)
+                if (!IsConnected && !loggedDisconnection)
                 {
+                    loggedDisconnection = true;
```
Comment (drwill-ms) [inline e2e/LongHaul/device/IotHub.cs:129]:
```suggestion'
               if (!IsConnected)
               {
                   if (!loggedDisconnection)
                   {
                       _logger.Trace($"Waiting for connection before sending telemetry", TraceSeverity.Warning);
                        loggedDisconnection = true;
                    }
                    continue;
                }
```

File: e2e/LongHaul/device/IotHub.cs [hunk 13]
```diff
@@ -158,15 +162,15 @@ public async Task ReportReadOnlyPropertiesAsync(CancellationToken ct)
...
                         { "TotalTelemetryMessagesSent", _totalTelemetryMessagesSent },
                     };
-                    await _deviceClient.UpdateReportedPropertiesAsync(reported ,ct).ConfigureAwait(false);
+                    await _deviceClient.UpdateReportedPropertiesAsync(reported, ct).ConfigureAwait(false);
 
                     ++_totalTwinUpdatesReported;
                     _logger.Metric(TotalTwinUpdatesReported, _totalTwinUpdatesReported);
                 }
-                else
+                else if(!loggedDisconnection)
```
Comment (drwill-ms) [inline e2e/LongHaul/device/IotHub.cs:170]:
```suggestion
                else if (!loggedDisconnection)
```

File: e2e/LongHaul/service/IotHub.cs [hunk 14]
```diff
@@ -111,6 +111,10 @@ await Task
                             {
                                 _logger.Trace($"Operations on [{deviceId}] have been canceled as the device goes offline.", TraceSeverity.Information);
                             }
+                            catch (Exception ex)
+                            {
+                                _logger.Trace($"Service app failed with exception {ex}", TraceSeverity.Error);
```
Comment (drwill-ms) [inline e2e/LongHaul/service/IotHub.cs:116]: Please add a `ToString()` override to our exception classes in iot hub (consider dps) to include ErrorCode, etc.

like:
Message
ErrorCode={}, StatusCode={}, IsTransient={}, TrackingId={}
CallStack

**Review by drwill-ms — APPROVED** [review]

File: iothub/device/src/Exceptions/IotHubClientException.cs [hunk 15]
```diff
@@ -98,7 +98,9 @@ public override void GetObjectData(SerializationInfo info, StreamingContext cont
         /// <inheritdoc/>
         public override string ToString()
         {
-            return $"{Message}\nError code: {ErrorCode}\n{StackTrace}n";
+            return $"Message: {Message}\n" +
```
Comment (drwill-ms) [inline iothub/device/src/Exceptions/IotHubClientException.cs:101]: I'd prefer we use a string builder here

File: iothub/device/src/Exceptions/IotHubClientException.cs [hunk 16]
```diff
@@ -98,7 +98,9 @@ public override void GetObjectData(SerializationInfo info, StreamingContext cont
...
+                   $"ErrorCode: {ErrorCode}, TrackingId: {TrackingId}, IsTransient: {IsTransient}\n" +
```
Comment (drwill-ms) [inline iothub/device/src/Exceptions/IotHubClientException.cs:102]: You can make tracking Id optional if it is not a real value

File: iothub/service/src/Exceptions/IotHubServiceException.cs [hunk 17]
```diff
@@ -104,6 +104,14 @@ public override void GetObjectData(SerializationInfo info, StreamingContext cont
             info.AddValue(TrackingIdValueSerializationStoreName, TrackingId);
         }
 
+        /// <inheritdoc/>
+        public override string ToString()
+        {
+            return $"Message: {Message}\n" +
```
Comment (drwill-ms) [inline iothub/service/src/Exceptions/IotHubServiceException.cs:110]: Suggested format would be:

$"{Message}\nError code={ErrorCode}, ...\n{StackTrace}"

**Pull request merged**